### ISSUE TITLE
Consider supporting smart crop

### ISSUE THREAD

**Issue by h2non** [jcupitt/libvips#317]
There're multiple implementations, but this one in JavaScript become popular based on canvas processing:
https://github.com/jwagner/smartcrop.js

A similar feature natively supported by `libvips` would be great!

**Comment by jcupitt** [comment]
Oh, looks nice. It should be fairly easy to implement with vips in Python / Ruby / whatever, let's leave this open as a reminder.

**Comment by jcupitt** [comment]
I had a quick go:

**Comment by h2non** [comment]
Oh, great first match! This is only applicable for human skin color, is it?

**Comment by jcupitt** [comment]
It looks for hues around 20 degrees, +/- a bit, I don't know how common that is.

**Comment by jcupitt** [comment]
I updated the gist, it does more looping in vips now and less in Python, down to about 200ms for the test image. It still needs tuning.

**Comment by jcupitt** [comment]
I fiddled a bit more, it works OK now. It still prefers small crops to large crops, which isn't quite right. I'm missing a trick somewhere.

**Comment by jcupitt** [comment]
Here's another interesting smart crop idea:

http://stackoverflow.com/a/1517178/894763

That one repeatedly chops low-entropy bits from the image until it's square, then shrinks the square part.

**Comment by jcupitt** [comment]
The stackoverflow idea in pyvips:

https://gist.github.com/jcupitt/c2792a9b22e4e2544147

it needs libvips git master, I added vips_hist_entropy()

**Comment by h2non** [comment]
Cool! Thank you for introducing that in `libvips`. Definitively I'll take a look in detail

**Comment by F21** [comment]
Are there any plans to introduce this directly into libvips?

**Comment by jcupitt** [comment]
Perhaps it could move into vips in 8.5. 

I think sharp now incorporates a version of this, it would be good to get an idea of how well it works on real-world image sets before baking it in.

**Comment by lovell** [comment]
Hello, here's my (annual?) update for this issue:

1: As John points out, `sharp` now provides a version of the "repeatedly slice off lowest entropy edge" algorithm as a "crop strategy", which can be accessed via:
```javascript
sharp(input)
  .resize(200, 200)
  .crop(sharp.strategy.entropy)
  ...
```

2: `smartcrop.js` is now adapter-based and can use [sharp+libvips as its backing library](https://github.com/jwagner/smartcrop-sharp).

3: I'm revisiting the [attention](https://github.com/lovell/attention) module, which involves the use of training data to determine its algorithm (artificial unintelligence), with the possibility of including it in `sharp` as an experimental option.

**Comment by jcupitt** [comment]
Hey @lovell, thanks for the update.

Do you have any feel for how well entropy cropping works on typical image sets? Is it good enough to be useful?

**Comment by jcupitt** [comment]
Also, interesting news on smartcrop having a sharp backend, I'll have a read.

**Comment by lovell** [comment]
> "Do you have any feel for how well entropy cropping works on typical image sets?"

Nothing scientific; in my experience it's almost always better than a simple centre-based crop though :)

The feedback on https://github.com/lovell/sharp/issues/295 has been that smartcrop.js is better than entropy alone, but that's what I'd expect.

Perhaps the best thing for me to do is convert https://gist.github.com/jcupitt/a27d449c280863619035 to C++, include it (experimentally) in sharp and gather more feedback (accuracy, performance) from users.

**Comment by rnanwani** [comment]
Since this is an experimental feature and would take a lot of time of testing on real images, can this be moved to a later milestone instead of 8.5?

**Comment by jcupitt** [comment]
Lovell and sharp have tested this pretty thoroughly in lovell/sharp#295, so I'm planning just to copy-paste that into vips. It should only be an hour's work (I think). 

If it turn out to be horrible, I'll bump the milestone.

**Comment by jcupitt** [comment]
Here's smartcrop in libvips:

https://github.com/jcupitt/libvips/blob/add-smartcrop/libvips/conversion/smartcrop.c

It's a slight variation on the one in sharp. I've leave it in the branch for a few days, if anyone has time to test it.

**Comment by jcupitt** [comment]
`vipsthumbnail` and `vips_thumbnail()` have `smartcrop` options now.

**Comment by puzrin** [comment]
https://www.flickr.com/photos/-nature-/11841147024/

![](https://c2.staticflickr.com/6/5525/11841147024_f7d0ec7dbe_o_d.jpg)

-> thumbnail 170x150, attention mode.

sharp's code still cut a lot from the right, instead of cut all from the left.

https://29a.ch/sandbox/2014/smartcrop/examples/testbed.html - this implementation gas no such issue.

**Comment by jcupitt** [comment]
That's a nice test image. With this command:

```
$ vipsthumbnail -s 170x150 --smartcrop attention owl.jpg 
```

I see:

![tn_owl](https://cloud.githubusercontent.com/assets/580843/23822278/e8d1c5a4-0641-11e7-9c9e-a12b65887951.jpg)

The lopsided cropping is down to the algorithm. Sharp (and now libvips) smartcrop, not by detecting objects, but by discarding non-objects. This is much faster, since you don't need to analyze the whole image, but it does mean it won't centre on objects.

**Comment by puzrin** [comment]
I did not dived into algorythm internals. IMHO:

1. Ordinary users (site visitors) do not care about details. They can say if they like result or not :). I think, result after `smartcrop.js` will be preferable in many cases.
2. Thumbnails are small (if you scale down before crop). I think, speed will not be a big problem in this case.
3. Even if algorythm is slow in your eyes, that's still much faster than JS implementation :)

In other words, i'd suggest to add one more smartcrop mode. Example above shows my reasons.

**Comment by jcupitt** [comment]
I think this would be a new algorithm. Let's make an enhancement so as to not delay 8.5 too much. I'll open a new issue.

**Comment by jcupitt** [comment]
New issue: https://github.com/jcupitt/libvips/issues/619

**Comment by jcupitt** [comment]
OK, merged to master. Thanks!

**Comment by jrochkind** [comment]
Has any smartcrop made it into a vips release yet? I'm on 8.5.9, and I don't _think_ I have it as an argument to `vips thumbnail` (at least it doens't show up in help screen).  I'd love this option, even a non-perfect one that is just usually better than centering.

**Comment by jcupitt** [comment]
Hello, yes, it's in 8.5. Try:

```
$ vipsthumbnail k2.jpg --size 100x100 --smartcrop attention 
```

It's not part of the `thumbnail` operator, it's a separate thing. Try:

```
$ vips smartcrop
extract an area from an image
usage:
   smartcrop input out width height [--option-name option-value ...]
where:
   input        - Input image, input VipsImage
   out          - Output image, output VipsImage
   width        - Width of extract area, input gint
			default: 1
			min: 1, max: 10000000
   height       - Height of extract area, input gint
			default: 1
			min: 1, max: 10000000
optional arguments:
   interesting  - How to measure interestingness, input VipsInteresting
			default: attention
			allowed: none, centre, entropy, attention
```

**Comment by jrochkind** [comment]
Ah, I see, thank you. I thought `vipsthumbnail` just did things that `vips thumbnail` did, didn't realize it sometimes incorporated other operations too, okay!

**Comment by galdolber** [comment]
Is there any way to have the smartcrop command return the x,y,w,h that resulted?

**Comment by jcupitt** [comment]
Sure, look at `xoffset` / `yoffset` in the output image.
